### PR TITLE
Add 96Boards WisTrio board support

### LINKED ISSUE #9333: Support for STM32 L1-series
I'm currently looking around for a small RTOS to use in a new project for a customer and Zephyr matches more or less perfectly what I need. My current problem is that the CPU used in the project is based on the STM32L1 series (STM32L152RE to be more exact) and Zephyr does not seem to support that at the moment. So I would like to request support for L1 to be added.

https://www.st.com/en/embedded-software/stm32cubel1.html

### PR DESCRIPTION
**Pull request by Mani-Sadhasivam**
This PR adds 96Boards WisTrio board support with STM32L1 SoC series and HAL.

Fixes #9333

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

### CODE REVIEW

**Comment by erwango** [comment]
@nashif, I'm surprised with the warning on Cube files license. It's usual BSD-3 clause as used in already in tree versions. Is there anything new on that front?

**Comment by Mani-Sadhasivam** [comment]
@erwango Do you have any idea about the Sanity check failure?

**Comment by erwango** [comment]
@Mani-Sadhasivam, please fix Licence issues for files that are not under ext/. For files under ext/, warning will be discarded.

**Comment by Mani-Sadhasivam** [comment]
@erwango Fixed the license for non ext/ files. But I'm not sure about the shippable failure for the testcase.

File: soc/arm/st_stm32/stm32l1/Kconfig.defconfig.stm32l151xb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+if SOC_STM32L151XB
+
+config SOC
+	string
+	default "stm32l151xb"
+
+config NUM_IRQS
+	int
+	default 32
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l1/Kconfig.defconfig.stm32l151xb]: Can you double check this value?
You're using IRQ num > 32 in dts, this should be causing the issue detected by shippable

**Review by dbkinder — APPROVED** [review]
Doc additions LGTM, thanks!

**Comment by erwango** [comment]
@Mani-Sadhasivam , can you rebase on top of https://github.com/zephyrproject-rtos/zephyr/pull/12024, this is GPIO driver refactoring, will spare you some code (bit late though ;-))

File: soc/arm/st_stm32/stm32l1/Kconfig.defconfig.stm32l151xb (hunk 1 above)
Comment (Mani-Sadhasivam) [reply soc/arm/st_stm32/stm32l1/Kconfig.defconfig.stm32l151xb]: Ah. It is 45. Updated in next revision!

**Comment by Mani-Sadhasivam** [comment]
@erwango I'd like to rebase once that PR gets merged. Or we can do the conversion as an incremental PR after merging this!

**Comment by erwango** [comment]
@Mani-Sadhasivam, this is now merged.  Please take it into account as of now, I prefer not merging code I know it is already obsolete. Besides, due to factorization, this will be less code to review :-)

**Comment by erwango** [comment]
@Mani-Sadhasivam, btw, and so you're not disappointed, there are significant changes on going (particularly on dt front), which might have impact on SoC porting and general and hence on this PR. These changes are meant to be integrated in 1.14 LTS.
I'm quite happy to see L1 port coming (thank you for that), but for coherency, I'd like STM32L1 port to be identical to other STM32 series, specially if merged in LTS. So I'll be more than happy to have this PR merged for 1.14, but this means that you'll have to integrate all these on going changes in a short period (and even though, I can't swear it'll will make it).

File: soc/arm/st_stm32/stm32l1/Kconfig.series [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+#
+# Copyright (c) 2019 Linaro Ltd.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+config SOC_SERIES_STM32L1X
+	bool "STM32L1x Series MCU"
+	select CPU_CORTEX_M3
+	select CPU_CORTEX_M_HAS_VTOR
```
Comment (tagunil) [inline soc/arm/st_stm32/stm32l1/Kconfig.series]: You don't need to select CPU_CORTEX_M_HAS_VTOR here because it's already selected by CPU_CORTEX_M3.

**Comment by nashif** [comment]
> @nashif, I'm surprised with the warning on Cube files license. It's usual BSD-3 clause as used in already in tree versions. Is there anything new on that front?

bsd-new is bsd-3 clause

File: drivers/interrupt_controller/exti_stm32.c [hunk 3]
```diff
@@ -43,6 +43,8 @@
 #define EXTI_LINES 24
 #elif defined(CONFIG_SOC_SERIES_STM32L0X)
 #define EXTI_LINES 30
+#elif defined(CONFIG_SOC_SERIES_STM32L1X)
+#define EXTI_LINES 40
```
[the remark this answers is not in the document]
Comment (Mani-Sadhasivam) [reply drivers/interrupt_controller/exti_stm32.c]: Oh yeah... It is 24. Will change it!

File: drivers/i2c/Kconfig.stm32 [hunk 4]
```diff
@@ -14,7 +14,7 @@ if I2C_STM32
 
 config I2C_STM32_V1
 	bool "STM32 V1 Driver (F1/F4X)"
```
[the remark this answers is not in the document]
Comment (Mani-Sadhasivam) [reply drivers/i2c/Kconfig.stm32]: Ack

File: soc/arm/st_stm32/stm32l1/Kconfig.series (hunk 2 above)
Comment (Mani-Sadhasivam) [reply soc/arm/st_stm32/stm32l1/Kconfig.series]: Ack

**Comment by Mani-Sadhasivam** [comment]
@erwango @tagunil Here is the changeset.

* Fixed EXTI lines
* Added IRQ defines for STM32L1X EXTI
* Adopted the GPIO support based on the recently merged STM32 GPIO refactoring PR.

File: drivers/interrupt_controller/exti_stm32.c [hunk 5]
```diff
@@ -711,6 +730,35 @@ static void __stm32_exti_connect_irqs(struct device *dev)
 		CONFIG_EXTI_STM32_EXTI15_4_IRQ_PRI,
 		__stm32_exti_isr_4_15, DEVICE_GET(exti_stm32),
 		0);
+#elif defined(CONFIG_SOC_SERIES_STM32L1X)
```
Comment (erwango) [inline drivers/interrupt_controller/exti_stm32.c]: Please rebase on top of https://github.com/zephyrproject-rtos/zephyr/pull/12463 for factorization

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/arm/st/l1/stm32l15XXB.dtsi [hunk 6]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright (c) 2019 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <mem.h>
+#include <st/l1/stm32l1.dtsi>
```
[the remark this answers is not in the document]
Comment (Mani-Sadhasivam) [reply dts/arm/st/l1/stm32l15XXB.dtsi]: I don't find any specific difference to add `stm32l151.dtsi` file, that's why I added `stm32l15XXB.dtsi` with common `stm32l1.dtsi`. Can you please help?

File: soc/arm/st_stm32/stm32l1/soc.h [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+ *   STM32L1X advanced ARM ® -based 32-bit MCUs
+ *
+ * Chapter 2.2: Memory organization
+ */
+
+
+#ifndef _STM32L1_SOC_H_
+#define _STM32L1_SOC_H_
+
+#define GPIO_REG_SIZE         0x400
```
[the remark this answers is not in the document]
Comment (Mani-Sadhasivam) [reply soc/arm/st_stm32/stm32l1/soc.h]: Okay

File: drivers/pinmux/stm32/pinmux_stm32l1x.h [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+#ifndef ZEPHYR_DRIVERS_PINMUX_STM32_PINMUX_STM32L1X_H_
+#define ZEPHYR_DRIVERS_PINMUX_STM32_PINMUX_STM32L1X_H_
+
+/**
+ * @file Header for STM32L1X pin multiplexing helper
+ */
+
+/* Port A */
+#define STM32L1X_PINMUX_FUNC_PA0_PWM2_CH1   \
```
[the remark this answers is not in the document]
Comment (Mani-Sadhasivam) [reply drivers/pinmux/stm32/pinmux_stm32l1x.h]: Most of them are copied from L4X. Will cross check this in next iteration.

File: boards/arm/96b_wistrio/96b_wistrio.yaml [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+name: 96boards WisTrio
+type: mcu
+arch: arm
+toolchain:
+  - zephyr
+  - gnuarmemb
+ram: 16
+flash: 128
+supported:
+  - gpio
```
[the remark this answers is not in the document]
Comment (Mani-Sadhasivam) [reply boards/arm/96b_wistrio/96b_wistrio.yaml]: Okay

File: soc/arm/st_stm32/stm32l1/soc.h [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+#include <stm32l1xx_ll_utils.h>
+#include <stm32l1xx_ll_bus.h>
+#include <stm32l1xx_ll_rcc.h>
+#endif /* CONFIG_CLOCK_CONTROL_STM32_CUBE */
+
+#ifdef CONFIG_I2C_STM32_V1
+#include <stm32l1xx_ll_i2c.h>
+#endif
+
+#ifdef CONFIG_SPI_STM32
```
[the remark this answers is not in the document]
Comment (Mani-Sadhasivam) [reply soc/arm/st_stm32/stm32l1/soc.h]: Okay

File: drivers/interrupt_controller/exti_stm32.c (hunk 5 above)
Comment (Mani-Sadhasivam) [reply drivers/interrupt_controller/exti_stm32.c]: Okay

File: dts/arm/st/l1/stm32l15XXB.dtsi (hunk 6 above)
[the remark this answers is not in the document]
Comment (erwango) [reply dts/arm/st/l1/stm32l15XXB.dtsi]: No problem, stm32l151.dtsi could be empty.
Aim is to have same hierarchy as on other series, which proved to scale easily when new SoCs have to be ported.

File: drivers/pinmux/stm32/pinmux_stm32l1x.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (erwango) [reply drivers/pinmux/stm32/pinmux_stm32l1x.h]: ok, txs

File: dts/arm/st/l1/stm32l15XXB.dtsi (hunk 6 above)
[the remark this answers is not in the document]
Comment (Mani-Sadhasivam) [reply dts/arm/st/l1/stm32l15XXB.dtsi]: Okay, cool. Added!

**Comment by erwango** [comment]
@Mani-Sadhasivam , fyi, #12463 has been merged.

**Comment by Mani-Sadhasivam** [comment]
@erwango I couldn't reproduce the Shippable failure. Is it due to commits reordering by github? Or I'm missing something?

File: soc/arm/st_stm32/stm32l1/soc.h [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+#include <stm32l1xx.h>
+
+/* ARM CMSIS definitions must be included before kernel_includes.h.
+ * Therefore, it is essential to include kernel_includes.h after including
+ * core SOC-specific headers.
+ */
+#include <kernel_includes.h>
+
+#include <stm32l1xx_ll_system.h>
+
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l1/soc.h]: Miss
```
#ifdef CONFIG_EXTI_STM32
#include <stm32l1xx_ll_exti.h>
#endif
```

Comment (erwango) [reply soc/arm/st_stm32/stm32l1/soc.h]: @Mani-Sadhasivam, By the way, can I ask you to have a cleaner split of your commits between wrt functionalities addition? I know this requires bit more work on your side, but for instance, it's harder to review when I2C changes are split among multiple commits.

File: soc/arm/st_stm32/stm32l1/Kconfig.defconfig.stm32l151xb [hunk 12]
```diff
@@ -15,4 +15,26 @@ config NUM_IRQS
 	int
 	default 45
 
+if GPIO_STM32
+
+config GPIO_STM32_PORTA
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l1/Kconfig.defconfig.stm32l151xb]: GPIOA, B and C are enabled in `st_stm32/common`

File: soc/arm/st_stm32/stm32l1/dts_fixup.h [hunk 13]
```diff
@@ -8,4 +8,20 @@
...
+#define DT_UART_STM32_USART_1_CLOCK_BUS	DT_ST_STM32_USART_40013800_CLOCK_BUS
+
+#define DT_UART_STM32_USART_2_BASE_ADDRESS	DT_ST_STM32_USART_40004400_BASE_ADDRESS
+#define DT_UART_STM32_USART_2_BAUD_RATE	DT_ST_STM32_USART_40004400_CURRENT_SPEED
+#define DT_UART_STM32_USART_2_IRQ_PRI	DT_ST_STM32_USART_40004400_IRQ_0_PRIORITY
+#define DT_UART_STM32_USART_2_NAME		DT_ST_STM32_USART_40004400_LABEL
+#define DT_USART_2_IRQ				DT_ST_STM32_USART_40004400_IRQ_0
+#define DT_UART_STM32_USART_2_CLOCK_BITS	DT_ST_STM32_USART_40004400_CLOCK_BITS
+#define DT_UART_STM32_USART_2_CLOCK_BUS	DT_ST_STM32_USART_40004400_CLOCK_BUS
+
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l1/dts_fixup.h]: Add USART3 since it is provided in dts

**Review by erwango — CHANGES_REQUESTED** [review]
Few comments otherwise sounds good.

File: dts/arm/st/l1/stm32l15XXB.dtsi [hunk 14]
```diff
@@ -0,0 +1,22 @@
+/*
```
[the remark this answers is not in the document]
Comment (Mani-Sadhasivam) [reply dts/arm/st/l1/stm32l15XXB.dtsi]: Done

File: drivers/clock_control/Kconfig.stm32 [hunk 15]
```diff
@@ -322,6 +322,27 @@ config CLOCK_STM32_PLL_DIVISOR
 
 endif # SOC_SERIES_STM32L0X
 
+if SOC_SERIES_STM32L1X
+
+config CLOCK_STM32_PLL_MULTIPLIER
```
[the remark this answers is not in the document]
Comment (Mani-Sadhasivam) [reply drivers/clock_control/Kconfig.stm32]: Okay

File: soc/arm/st_stm32/stm32l1/dts_fixup.h [hunk 16]
```diff
@@ -78,4 +78,14 @@
 #define DT_UART_STM32_USART_2_CLOCK_BITS	DT_ST_STM32_USART_40004400_CLOCK_BITS
 #define DT_UART_STM32_USART_2_CLOCK_BUS	DT_ST_STM32_USART_40004400_CLOCK_BUS
 
+#define DT_I2C_1_BASE_ADDRESS               DT_ST_STM32_I2C_V1_40005400_BASE_ADDRESS
```
[the remark this answers is not in the document]
Comment (Mani-Sadhasivam) [reply soc/arm/st_stm32/stm32l1/dts_fixup.h]: Done

File: soc/arm/st_stm32/stm32l1/soc.h [hunk 17]
```diff
@@ -47,6 +47,10 @@
 #include <stm32l1xx_ll_exti.h>
 #endif
 
+#ifdef CONFIG_I2C_STM32_V1
```
[the remark this answers is not in the document]
Comment (Mani-Sadhasivam) [reply soc/arm/st_stm32/stm32l1/soc.h]: OKay

File: boards/arm/96b_wistrio/dts_fixup.h [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * This file is a temporary workaround for mapping of the generated information
+ * to the current driver definitions.  This will be removed when the drivers
+ * are modified to handle the generated information, or the mapping of
+ * generated data matches the driver definitions.
+ */
+
+#define DT_LIS3DH_DEV_NAME			DT_ST_STM32_I2C_V1_40005400_ST_LIS3DH_32_LABEL
```
[the remark this answers is not in the document]
Comment (Mani-Sadhasivam) [reply boards/arm/96b_wistrio/dts_fixup.h]: Okay

File: soc/arm/st_stm32/stm32l1/Kconfig.defconfig.stm32l151xb (hunk 12 above)
Comment (Mani-Sadhasivam) [reply soc/arm/st_stm32/stm32l1/Kconfig.defconfig.stm32l151xb]: Okay

File: soc/arm/st_stm32/stm32l1/dts_fixup.h (hunk 13 above)
Comment (Mani-Sadhasivam) [reply soc/arm/st_stm32/stm32l1/dts_fixup.h]: Okay

File: soc/arm/st_stm32/stm32l1/Kconfig.series [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+# Copyright (c) 2019 Linaro Ltd.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+config SOC_SERIES_STM32L1X
+	bool "STM32L1x Series MCU"
+	select CPU_CORTEX_M3
+	select SOC_FAMILY_STM32
+	select SYS_POWER_LOW_POWER_STATE_SUPPORTED
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l1/Kconfig.series]: Has been renamed into SYS_POWER_LOW_POWER_STATES_SUPPORTED

**Review by erwango — CHANGES_REQUESTED** [review]
Last update otherwise it should be fine

File: soc/arm/st_stm32/stm32l1/Kconfig.series (hunk 19 above)
Comment (Mani-Sadhasivam) [reply soc/arm/st_stm32/stm32l1/Kconfig.series]: Done

**Review by erwango — APPROVED** [review]

**Comment by Mani-Sadhasivam** [comment]
@erwango I think this PR won't make it for v1.14 LTS?

**Comment by erwango** [comment]
> @erwango I think this PR won't make it for v1.14 LTS?

Indeed, it won't. But it will be soon merged once 1.15 window is open.

**Comment by galak** [comment]
Some minor things that need fixing up.

**Comment by frantony** [comment]
@Mani-Sadhasivam 

Something is wrong with ```_ClearFaults```:

```shell
$ sudo apt-get install -y wget python3 python3-yaml python3-pyelftools make cmake gperf git
$ cd ~
$ wget https://github.com/zephyrproject-rtos/sdk-ng/releases/download/v0.10.0/zephyr-sdk-0.10.0-setup.run
$ mkdir zephyr-sdk
$ cd zephyr-sdk/
$ export TMPDIR=/tmp
$ sh ../zephyr-sdk-0.10.0-setup.run -- -d $(pwd)

$ cd ~
$ git clone https://github.com/Mani-Sadhasivam/zephyr
$ cd zephyr
$ git checkout 96b_wistrio
$ git describe
v1.8.99-15325-ga3d76ee3ef
$ export ZEPHYR_TOOLCHAIN_VARIANT=zephyr
$ export ZEPHYR_SDK_INSTALL_DIR=~/zephyr-sdk
$ source zephyr-env.sh
$ cd samples/hello_world
$ rm -rf build && mkdir build && cd build
$ cmake -DBOARD=96b_wistrio ..
$ make

  ...

[ 94%] Building C object zephyr/CMakeFiles/zephyr_prebuilt.dir/misc/empty_file.c.obj
[ 95%] Linking C executable zephyr_prebuilt.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       11212 B       128 KB      8.55%
            SRAM:        4116 B        16 KB     25.12%
        IDT_LIST:         120 B         2 KB      5.86%
/home/antony/zephyr-sdk/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/../../../../arm-zephyr-eabi/bin/ld: libzephyr.a(soc.c.obj): in function `stm32l1_init':
/home/antony/zephyr/soc/arm/st_stm32/stm32l1/soc.c:35: undefined reference to `_ClearFaults'
collect2: error: ld returned 1 exit status
make[2]: *** [zephyr/CMakeFiles/zephyr_prebuilt.dir/build.make:96: zephyr/zephyr_prebuilt.elf] Error 1
make[1]: *** [CMakeFiles/Makefile2:455: zephyr/CMakeFiles/zephyr_prebuilt.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

**Comment by galak** [comment]
> Something is wrong with `_ClearFaults`:

it got renamed to `z_clearfaults` if I'm not mistaken.

**Comment by frantony** [comment]
@galak 

> > Something is wrong with _ClearFaults:
>
> it got renamed to z_clearfaults if I'm not mistaken.

You are right! Thanks! It got renamed in fd42bf7443cb2f36619be0eab7aa378c30f49da7.

**Comment by Mani-Sadhasivam** [comment]
@galak Fixed the issues!

**Review by galak — APPROVED** [review]

**Pull request merged**